File: src/main/java/com/artillexstudios/axminions/listeners/WorldListener.java
package com.artillexstudios.axminions.listeners;

import com.artillexstudios.axapi.scheduler.Scheduler;
import com.artillexstudios.axapi.utils.logging.LogUtils;
import com.artillexstudios.axminions.AxMinionsPlugin;
import com.artillexstudios.axminions.config.Config;
import com.artillexstudios.axminions.minions.MinionArea;
import com.artillexstudios.axminions.minions.MinionWorldCache;
import org.bukkit.Bukkit;
import org.bukkit.Chunk;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.world.WorldLoadEvent;
import org.bukkit.event.world.WorldUnloadEvent;

public final class WorldListener implements Listener {

    @EventHandler
    public void onWorldLoadEvent(WorldLoadEvent event) {
        MinionArea area = MinionWorldCache.loadArea(event.getWorld());
        AxMinionsPlugin.instance().handler().loadMinions(event.getWorld()).toCompletableFuture()
                .thenAccept(loaded -> {
                    if (Config.debug) {
                        LogUtils.debug("Loaded {} minions in world {} in {} ms!", loaded.firstInt(), event.getWorld().getName(), loaded.secondLong() / 1_000_000);
                    }

                    if (area == null) {
                        return;
                    }

                    Scheduler.get().run(() -> {
                        if (Bukkit.getWorld(event.getWorld().getUID()) == null) {
                            return;
                        }

                        for (Chunk chunk : event.getWorld().getLoadedChunks()) {
                            area.startTicking(chunk);
                        }
                    });
                });
    }

    @EventHandler
    public void onWorldUnloadEvent(WorldUnloadEvent event) {
        MinionWorldCache.clear(event.getWorld());
        MinionWorldCache.remove(event.getWorld());
    }
}

File: common/src/main/kotlin/com/artillexstudios/axminions/listeners/WorldListener.kt
package com.artillexstudios.axminions.listeners

import com.artillexstudios.axminions.api.minions.miniontype.MinionTypes
import com.artillexstudios.axminions.api.utils.fastFor
import com.artillexstudios.axminions.minions.Minions
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.world.WorldLoadEvent
import org.bukkit.event.world.WorldUnloadEvent

class WorldListener : Listener {

    @EventHandler
    fun onWorldLoadEvent(event: WorldLoadEvent) {
        MinionTypes.loadForWorld(event.world)

        event.world.loadedChunks.fastFor {
            Minions.startTicking(it)
        }
    }

    @EventHandler
    fun onWorldUnload(event: WorldUnloadEvent) {
        val worldUUID = event.world.uid
        Minions.getMinions().fastFor {
            if (it.getLocation().world?.uid == worldUUID) {
                Minions.remove(it)
            }
        }
    }
}